server/ast/expr.go

```diff
@@ -100,25 +101,38 @@ func nodeExpr(node tree.Expr) (vitess.Expr, error) {
 		return nil, fmt.Errorf("ANNOTATE_TYPE is not yet supported")
 	case *tree.Array:
 		//TODO: right now, this only works with boolean array values for the sake of demonstration
-		var gmsExpr sql.Expression
+		var sqlChildren []sql.Expression
+		var unresolvedChildren []vitess.Expr
+		var unresolvedIndexes []int
 		if len(node.Exprs) == 0 {
 			if node.ResolvedType().Family() == types.ArrayFamily && node.ResolvedType().ArrayContents().Family() == types.BoolFamily {
-				gmsExpr = expression.NewLiteral([]bool{}, pgtypes.BoolArray)
+				sqlChildren = []sql.Expression{expression.NewLiteral([]bool{}, pgtypes.BoolArray)}
 			} else {
 				return nil, fmt.Errorf("arrays are generally not yet supported")
 			}
 		} else {
-			vals := make([]bool, len(node.Exprs))
-			for i, arrayExpr := range node.Exprs {
+			for _, arrayExpr := range node.Exprs {
 				if arrayVal, ok := arrayExpr.(*tree.DBool); ok && arrayVal != nil {
-					vals[i] = bool(*arrayVal)
+					sqlChildren = append(sqlChildren, expression.NewLiteral(bool(*arrayVal), pgtypes.Bool))
```

Comment:
Should the `*tree.DBool` be recursively resolved as a separate `InjectedExpression` with no children? Otherwise you'd have to switch on all your future custom types, and vice-versa

Reply:
Yeah it will be resolved recursively. We already have the setup for it (`nodeExpr`), but I didn't bother since this is just a demo implementation for now.